### PR TITLE
Record lifetime constraints inside generator interior

### LINKED ISSUE #64552: Strange error when iterating over BTreeMap in async fn
```rust
use std::collections::{BTreeMap, HashMap};
use std::sync::Arc;

fn needs_send<T: Send>(_val: T) {}
async fn async_fn_a(_num: u32) {}
async fn async_fn_b(map: Arc<BTreeMap<u32, &'static u32>>) {
    for (_i, v) in &*map {
        async_fn_a(**v).await;
    }
}
async fn async_fn_c(map: Arc<HashMap<u32, &'static u32>>) {
    for (_i, v) in &*map {
        async_fn_a(**v).await;
    }
}

fn main() {
    // this works...
    let map: Arc<HashMap<u32, &'static u32>> = Arc::new(HashMap::new());
    needs_send(async_fn_c(map.clone()));
    
    // but this doesn't
    let map: Arc<BTreeMap<u32, &'static u32>> = Arc::new(BTreeMap::new());
    needs_send(async_fn_b(map.clone()));
}
```
[(playground)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=12ee6a660148c7ba94503fc301a443a0)

```
error: implementation of `std::marker::Send` is not general enough
  --> src/main.rs:24:5
   |
24 |     needs_send(async_fn_b(map.clone()));
   |     ^^^^^^^^^^
   |
   = note: `std::marker::Send` would have to be implemented for the type `alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Immut<'0>, u32, &'1 u32, alloc::collections::btree::node::marker::Leaf>`, for any two lifetimes `'0` and `'1`
   = note: but `std::marker::Send` is actually implemented for the type `alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Immut<'2>, u32, &u32, alloc::collections::btree::node::marker::Leaf>`, for some specific lifetime `'2`
```

### PR DESCRIPTION
**Pull request by csmoe**
Closes #64552 

r? @Aaron1011

### CODE REVIEW

**Comment by tmandry** [comment]
Ping from triage. @csmoe, what's the status of this PR?

File: src/librustc_infer/infer/region_constraints/mod.rs [hunk 1]
```diff
@@ -137,14 +137,23 @@ pub enum Constraint<'tcx> {
...
         match self {
             Constraint::VarSubVar(_, _) => false,
             Constraint::VarSubReg(_, r) | Constraint::RegSubVar(r, _) => r.is_placeholder(),
             Constraint::RegSubReg(r, s) => r.is_placeholder() || s.is_placeholder(),
         }
     }
+
+    pub fn to_region_outlives_predicate(self) -> OutlivesPredicate<RegionKind, RegionKind> {
+        match self {
+            Self::VarSubVar(a, b) => OutlivesPredicate(RegionKind::ReVar(a), RegionKind::ReVar(b)),
```
Comment (csmoe) [inline src/librustc_infer/infer/region_constraints/mod.rs]: `RegionOutlivesPredicate<'tcx> = OutlivesPredicate<Region<'tcx>, Region<'tcx>>`, but I got `OutlivesPredicate<RegionKind<'tcx>>` here, struck by how to convert it into `RegionOutlivesPredicate<'tcx>`(cannot return [Region](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/type.Region.html) which refers to a local RegionKind.

**Comment by Aaron1011** [comment]
I think it would also be a good idea to record the `RegionVariableOrigin` for each region that we record a predicate for. This will allow us to generate more precise error messages during borrowck.

File: src/librustc/ty/context.rs [hunk 2]
```diff
@@ -2338,8 +2338,12 @@ impl<'tcx> TyCtxt<'tcx> {
     }
 
     #[inline]
-    pub fn mk_generator_witness(self, types: ty::Binder<&'tcx List<Ty<'tcx>>>) -> Ty<'tcx> {
-        self.mk_ty(GeneratorWitness(types))
+    pub fn mk_generator_witness(
+        self,
+        types: ty::Binder<&'tcx List<Ty<'tcx>>>,
+        region_outlives: &'tcx List<Predicate<'tcx>>,
```
Comment (csmoe) [inline src/librustc/ty/context.rs]: @Aaron1011 Should `List<RegionOutlivesPredicate>` be wrapped inside `Binder` like types in witness?

Comment (Aaron1011) [reply src/librustc/ty/context.rs]: @csmoe: I think so (assuming it also contains late-bound regions that were 'anonymized' from the original regions).

**Comment by benschulz** [comment]
@csmoe, in case you're unaware, the Github account of whomever you originally asked for a review is gone.

**Comment by Aaron1011** [comment]
@csmoe: Note that I'm not a part of any Rust team, so you'll probably want someone else to review this as well :smile:

**Comment by Dylan-DPC** [comment]
@csmoe can you rebase and resolve the conflicts? After that i can get this reviewed

**Comment by csmoe** [comment]
panic during `relation`, will investigate more.

**Comment by Dylan-DPC** [comment]
@csmoe any updates?

**Comment by csmoe** [comment]
@Dylan-DPC working on the panic.

**Comment by crlf0710** [comment]
This needs a rebase.

**Comment by Muirrum** [comment]
@csmoe This is a triage bump.

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Thanks

**Pull request closed without merging**